src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs

```diff
@@ -520,14 +520,40 @@ private static bool IsAllDigits(string filterText)
                 CompletionItem bestFilterMatch,
                 SnapshotPoint caretPosition,
                 CompletionHelper completionHelper,
-                CompletionFilterReason reason)
+                CompletionFilterReason filterReason)
             {
-                if (bestFilterMatch == null || model.UseSuggestionMode)
+                var itemViewSpan = model.GetViewBufferSpan(bestFilterMatch.Span);
+                var fullFilterText = model.GetCurrentTextInSnapshot(itemViewSpan, caretPosition.Snapshot, endPoint: null);
+                var textSpan = itemViewSpan.TextSpan;
+
+                // Switch to soft selection, if user moved caret to the start of a non-empty filter span.
+                // This prevents commiting if user types a commit character at this position later, but 
+                // still has the list if user types filter character
+                // i.e. blah| -> |blah -> !|blah
+                // We want the filter span non-empty because we still want hard selection in the following case:
+                //
+                //  A a = new |
+                if (caretPosition == textSpan.Start && textSpan.Length > 0)
```

Comment:
this was pulled up from below.  so if someone calls into your new entrypoint, they won't hit this.  Will that lead to observably different behavior?  If not, why not?

Reply:
In the new completion, we cannot handle caret position. We ae called with an async method and cannot guarantee that the caret position is actual. Therefore, the new Roslyn implementation cannot be responsible for checking the condition.

---
In reply to: [232509308](https://github.com/dotnet/roslyn/pull/29016#discussion_r232509308) [](ancestors = 232509308)

Reply:
This seems somewhat broken about the new completion.  For filtering, it's necessary to know where the user is (so we can filter up to thta point).  Is the editor not passing us in the snapshot point for the caret position?  That's how the old system worked.

Reply:
> Roslyn is called for Update(=Filter) in an async method. The caret position is not available here. 

That is a problem.  The caretposition (an immutable snapshotpoint) shoudl be captured when the work is kicked off and should be passed through to here.

>  We just cannot make those checks on the Roslyn side in the new completion.

This seems like a problem.  Why does the 'new completion' system not have the ability to pass this along?  This is why we were capturing the caret position in roslyn in the first place on the UI thread to pass along to the BG to compute the model.  It sounds like somehow this requirement was not properly convyed to the editor to include in their API.

Reply:
I'd say that Editor should make those checks on their side. They can make it once for all item managers. Then, they should not provide us caretPosition. I do not see other purposes of caretPosition in filter scenarios. @amadeusw to confirm. 

---
In reply to: [233250482](https://github.com/dotnet/roslyn/pull/29016#discussion_r233250482) [](ancestors = 233250482)